### ISSUE TITLE
License Issue Since PuLP 2.2

### ISSUE THREAD

**Issue by s-j-v-zelst** [coin-or/pulp#394]
Hi,

Thanks for all your efforts in creating pulp, it is a very useful library for us, i.e. we are using pulp in our software project.
However, we are not able to update to a version >= 2.2, because of PuLP's dependency on amply.

Since amply is licensed under the Eclipse 1.0 license, so should any derivative work be (including pulp).
Hence, even though pulp's license is very permissive (whereas it actually should be Eclipse 1.0), we can still not use it as we transitively are loading amply which is licensed with Eclipse 1.0.
Sadly, the eclipse license is clashing with the license of our software solution.

I was wondering whether you guys are aware of this, and, whether you have any plans to i) update the pulp license to Eclipse 1.0 and/or somehow obtaining amply under a different license.

all the best,

bas.

**Comment by pchtsp** [comment]
Good day! 

I'm not sure about the license limitations and implications, since I'm not an expert. Having said that, my understanding is that the Eclipse derivative work refers to modifications or improvements of the software (not the *use* of the software from another software as is the case with PuLP and amply).

A very brief extract from [the wikipedia page](https://en.wikipedia.org/wiki/Eclipse_Public_License#Derivative_works):

> Linking to code (for example to a library) licensed under EPL automatically does not mean that your program is a derivative work

@tkralphs @stumitchell can correct me or complement this answer better. Maybe it's a good opportunity to choose a standard licensing for PuLP, actually. What do you think?

**Comment by s-j-v-zelst** [comment]
Hi @pchtsp; yes, it depends a bit on the definition of 'derivative work'.
My personal interpretation is that a work is a derivative of another work, if it cannot 'function' without the other component, then again, this is my personal interpretation.
In general, however, different licenses interpret this differently.

In any case, regardless of the possibility to relicense PuLP itself, I guess that, since we use GPL for our library, the use of amply still results in a 'clash' for our specific purpose, regardless of relicensing of PuLP (i.e., amply would remain under Eclipse).

**Comment by stumitchell** [comment]
If you are redistributing pulp, hence why you are concerned about
licencing. Just remove the amply files and licence.

At some point we want to split amply out into a separate project which has
a dependency on pulp and remove it from the pulp project.

Stu

**Comment by s-j-v-zelst** [comment]
Hi @stumitchell,

We are redistributing amply transitively through our dependency on PuLP...
Hence, we cannot remove the files.

see also: https://www.eclipse.org/legal/eplfaq.php#GPLCOMPATIBLE
whereas in my view, our work is a derivative work of PulP and amply.

Or I am missing out on something (could be I am not an expert in the matter)?

cheers,

bas.

**Comment by tkralphs** [comment]
There are a the issues raised in this thread and I wanted to clarify a few points (caveat: I'm not a lawyer and this is not legal advice).

- A derivative work is essentially some modification of an existing work. If I take a book you wrote and just change a few paragraphs, it's a derivative work. I obviously don't own the modified version. In general it would difficult for an older work  be derivative of a newer work, so PuLP  would not be a derivative of amply. You could combine two existing works in a way that the combination would be derivative of both of those works and then their respective licenses would have to be compatible and allow for the combination to be distributed under a single license somehow. 
- On the other hand, I would guess that amply _is_ a derivative of PuLP because it was split off from it. In general, this would mean that it should have the same license as PuLP, but the _owner_ of a work can re-license derivatives any way they choose. As an owner, I can even re-license the original work. I'm not actually sure who owns the amply copyright, but I guess it was developed under @stumitchell and he can re-license it. It's not clear to me, though, why it was re-licensed and if this was a conscious decision. Seems like keeping it under the same license as PuLP would make sense. But you definitely cannot strip the license out when redistributing it. That would be either re-licensing or distributing without a license neither of which is legal for a non-owner. 
- Having a dependency on another Python package is not the same as re-distributing it. When your project depends on another Python package, it is the user of the package who is obtaining the source and installing it. They obtain that source on their own by running the `pip` command. The entity that is redistributing the code in the dependency is PyPi and they are bound by the license. They can only redistribute if the license allows. When someone uploads a package to Pypi, they are implicitly (or explicitly, not sure) agreeing that the package can be legally redistributed according to the license. There can't really be a license issue arising by virtue of package dependence in Python. The exception would be if you incorporate the source code from another package into your package and redistribute the combination.  _Then_ you do have a derivative work and license issues kick in. 
- Issues arising from one code "linking" to another basically only arise with the GPL (and perhaps some other similar licenses), which makes some stronger legal requirements. These have never been tested in court to my knowledge, so it's all interpretation (and you will find many interpretations if you Google). Also, "linking" is used in the highly technical sense, not in the looser way it would be used in plain English. The Free Software Foundation claims that if your code "links" to a dynamically linked library that is under the GPL, then your code should be under the GPL because it becomes a derivative work. You could then perhaps go so far as to say that using the API of a GPL'd code in your code means your code should be under the GPL (though whether APIs are copyrightable is a whole other legal rabbit hole). In terms of Python packages, though, merely interpreting Python code is not "linking" so a pure Python package under the GPL would be no problem. The situation with C extensions is tricky, though.

In general, the situation with a pure Python package that you own and distribute is pretty simple and there shouldn't be any license issues. But you should of course consult a real intellectual property lawyer. If I were one, the bill for what I just wrote would be substantial :).

**Comment by s-j-v-zelst** [comment]
Hi @tkralphs,

Thanks for clarifying some points.
I agree that my initial statement about the license of pulp was wrong, i.e., it is fine that it is licensed under a different license as amply; even though it is depending on it.

For your last two points and terminology in general; I think this primarily depends on the license's / human interpretation of the matter...
When googling about these things, one tends to find a variety of 'views'.
For example, the first answer of https://opensource.stackexchange.com/questions/1579/when-is-my-project-a-derivative-work-in-an-interpreted-language contradicts your points on 'depending' on other libraries through pip.

However, let me rephrase this 'issue' to avoid a lengthy discussion on licensing (although I enjoy such discussions), i.e., none of us is a lawyer in the end (I am a scientist co-developing a simple python library).

Since you guys also drive the development of amply, are you fine if we use the latest version of pulp and redistribute our software under GPL?
In the end, if you guys are fine with this; let's just be pragmatic :-)

**Comment by pchtsp** [comment]
From my side: I'm fine with it.

On a related note, @stumitchell you mentioned taking out amply altogether: I thought the idea was to keep the reference inside PuLP but take out the code to a new package. But what you say makes sense.
I'll probably just take out the few references to amply that are left in PuLP and add a link to amply on the [plugins page](https://coin-or.github.io/pulp/plugins/index.html) of the docs.

If someone has a good way of adding a "deprecated warning" so it only appears when someone imports amply from PuLP, that would be great to know.

**Comment by stumitchell** [comment]
TLDR I am happy for you to redistribute pulp under the GPL

@pchtsp I would just make a breaking change to pulp and release a version 3 (perhaps with any other api changes you think are needed)

@s-j-v-zelst here is the issue where we moved amply out of pulp https://github.com/coin-or/pulp/issues/246 I had forgotten that we had done this already, TLDR pulp doesn't need amply to function even though it is a dependency, so I simply suggested removing the files as I had assumed that you were redistributing a copy of pulp bundled with your project (rather than having a dependency using pypi) and as such you just needed to remove all pypi files to avoid any issues.

But as you say, I am happy for you to redistribute pulp under the GPL, and its license is designed to be as unrestrictive as possible.

EPL1.0 is I think the recommendation by coin-or (I think @tkralphs) which is why Amply was re licensed to it (it was initially covered by the pulp licence which allows re-licencing). In Fact the previous versions you are loading contain the amply files but under the pulp license

I hope this isn't more confusing

**Comment by tkralphs** [comment]
@s-j-v-zelst Thanks for pointing out that link. That was some good food for thought. I agree with everything that was said there and perhaps I was a little too black and white. If you create a small extension of an existing Python package and distribute the extension as a completely separate package, requiring the user to install the original package using pip, I could see the argument that the combination of the original and the extension is a derivative work of the original. However, this strikes as something that if tested in court would keep lawyers employed for years just as the Oracle/Google suit did and it's not clear which way it would come out. The analogous situation for something like a book would seem to be that you write an extra chapter or supplement to an existing book and then sell that chapter completely separately. Although the chapter wouldn't make sense without the original book (it _depends_ on the original), I think it would be very hard to argue that the supplement is infringing. Of course, if you bind it up as a single volume and try to sell it, then there is little question. A lot of copyright law is not really a good fit for software, but it's what we have. I would still bet that a separately distributed extension of a Python package would not be treated as a derivative if it used an unmodified version of the original. But anyway... we can leave it to the lawyers.

@stumitchell I had forgotten that the MIT license does allow sublicensing, so technically, yes, I suppose anyone _could_ redistribute PuLP under the GPL, with or without permission. But given that PuLP is already under the MIT license, which is GLP-compatible and less restrictive, I don't really see any scenario in which this would be needed or desirable. If copies of PuLP are circulating under two different licenses and those versions diverge (seems likely), it wouldn't be possible to bring them back together. Suppose someone fixes some bugs or extends the PuLP version that is under the GPL. The changes couldn't be merged back into the original.

It seems what raised the original concern was the re-licensing of amply, which @stumitchell is not actually a dependency of PuLP and is in any case available under its original MIT license. So I guess the original reason for the re-licensing is not an issue. 

Just to clarify, amply itself will _not_ be able to be sublicensed once under the EPL. @stumitchell, there is no reason for amply to be re-licensed from the COIN-OR side. The MIT license is certainly a very convenient and desirable open source license. Although most COIN-OR code is under the EPL, the MIT license is EPL-compatible. Of course, it is up to you if you want amply to be more restricted in terms of license. amply is currently not even part of COIN-OR.

**Comment by s-j-v-zelst** [comment]
Hi guys;

Thanks for all the fast responses and the positive attitude towards integrating pulp!

cheers,

Bas.
